contrib/kafka/filters/network/source/mesh/command_handlers/produce.cc

```diff
@@ -0,0 +1,116 @@
+#include "contrib/kafka/filters/network/source/mesh/command_handlers/produce.h"
+
+#include "contrib/kafka/filters/network/source/external/responses.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace Kafka {
+namespace Mesh {
+
+constexpr static int16_t NO_ERROR = 0;
+
+ProduceRequestHolder::ProduceRequestHolder(AbstractRequestListener& filter,
+                                           UpstreamKafkaFacade& kafka_facade,
+                                           const std::shared_ptr<Request<ProduceRequest>> request)
+    : ProduceRequestHolder{filter, kafka_facade, PlaceholderRecordExtractor{}, request} {};
+
+ProduceRequestHolder::ProduceRequestHolder(AbstractRequestListener& filter,
+                                           UpstreamKafkaFacade& kafka_facade,
+                                           const RecordExtractor& record_extractor,
+                                           const std::shared_ptr<Request<ProduceRequest>> request)
+    : BaseInFlightRequest{filter}, kafka_facade_{kafka_facade}, request_{request} {
+  outbound_records_ = record_extractor.extractRecords(request_->data_.topics_);
+  expected_responses_ = outbound_records_.size();
+}
+
+void ProduceRequestHolder::startProcessing() {
+  // Main part of the proxy: for each outbound record we get the appropriate sink (effectively a
+  // facade for upstream Kafka cluster), and send the record to it.
+  for (const auto& outbound_record : outbound_records_) {
+    KafkaProducer& producer = kafka_facade_.getProducerForTopic(outbound_record.topic_);
+    // We need to provide our object as first argument, as we will want to be notified when the
+    // delivery finishes.
+    producer.send(shared_from_this(), outbound_record.topic_, outbound_record.partition_,
+                  outbound_record.key_, outbound_record.value_);
+  }
+  // Corner case handling:
+  // If we ever receive produce request without records, we need to notify the filter we are ready,
+  // because otherwise no notification will ever come from the real Kafka producer.
+  if (finished()) {
+    notifyFilter();
+  }
+}
+
+bool ProduceRequestHolder::finished() const { return 0 == expected_responses_; }
+
+// Find a record that matches provided delivery confirmation coming from Kafka producer.
+// If all the records got their delivery data filled in, we are done, and can notify the origin
+// filter.
+bool ProduceRequestHolder::accept(const DeliveryMemento& memento) {
```

Comment:
[future]
this method gets invoked thru future [RichKafkaProducer implementation](https://github.com/envoyproxy/envoy/pull/11936/files#diff-cf7872d08318648e46c048d2bb615fb4bfc6929386c2f2c4367be65d5936939aR122) going thru the dispatcher to handle [librdkafka delivery confirmation](https://github.com/envoyproxy/envoy/pull/11936/files#diff-cf7872d08318648e46c048d2bb615fb4bfc6929386c2f2c4367be65d5936939aR110)

TLDR: librdkafka callback (dedicated thread) ~> dispatcher ~> RichKafkaProducer (envoy worker thread) -> PRH::accept